### PR TITLE
Parallelize the top level nodes iteration

### PR DESCRIPTION
**Pull request by gskchaitanya**

### CODE REVIEW

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java [hunk 1]
```diff
@@ -102,6 +109,12 @@
   //<set ID, <ts hash, ts wrapper>>
   private Map<String, Map<Long, SeriesWrapper>> partials = Maps.newConcurrentMap();
   private List<byte[]> serialized_results = Lists.newArrayList();
+  
+  /** Used for the parallel stream*/
+  private ForkJoinPool customThreadPool = new ForkJoinPool(8);
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: need to static it so we don't create it each time.

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java [hunk 2]
```diff
@@ -241,23 +254,30 @@ public Object call(final ArrayList<TimeSeriesStringId> ids)
...
+                      spec_start,
+                      spec_end,
+                      result);
+                } catch (Exception e) {
+                  LOG.error("Failed to serialize ts: " + series, e);
+                  throw new QueryExecutionException("Unexpected exception "
+                      + "serializing ts: " + series, 0, e);
+                }
+                if (pair.getValue().types().contains(StatusType.TYPE)
+                    && pair.getValue().id() instanceof BaseTimeSeriesByteId) {
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: won't happen that it's a byte ID series, always string.
BUT we need to sync on the namespace list.

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java [hunk 3]
```diff
@@ -509,84 +529,89 @@ private void serializeSeries(
...
-          if (value == null) {
-            continue;
-          }
-          
-          if (iterator.getType() == NumericType.TYPE) {
-            if (writeNumeric((TimeSeriesValue<NumericType>) value, options, 
-                  iterator, json, result, start, end, wrote_values)) {
-              wrote_values = true;
+
+      lock.writeLock().lock(); // since json is not thread safe and we need to form the json in order
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: lets do a null check and have the lock null if we're not parallelling.

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java [hunk 4]
```diff
@@ -241,23 +254,30 @@ public Object call(final ArrayList<TimeSeriesStringId> ids)
...
+                } catch (Exception e) {
+                  LOG.error("Failed to serialize ts: " + series, e);
+                  throw new QueryExecutionException("Unexpected exception "
+                      + "serializing ts: " + series, 0, e);
+                }
+                if (pair.getValue().types().contains(StatusType.TYPE)
+                    && pair.getValue().id() instanceof BaseTimeSeriesByteId) {
+                  BaseTimeSeriesStringId bid = (BaseTimeSeriesStringId) pair.getValue().id();
+                  namespace.append(bid.namespace());
+                  wasStatus.getAndSet(true);
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: wasStatus needs to be volatile now.

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java [hunk 5]
```diff
@@ -226,12 +239,12 @@ public Object call(final ArrayList<TimeSeriesStringId> ids)
...
-          boolean wasStatus = false;
-          boolean wasEvent = false;
-          String namespace = null;
-          if (opts.getParallelThreshold() > 0 &&
-              result.timeSeries().size() > opts.getParallelThreshold()) {
+          AtomicInteger ai = new AtomicInteger(0);
+          AtomicBoolean wasStatus = new AtomicBoolean(false);
+          StringBuilder namespace = new StringBuilder();
+          Collection<QueryNode> topNExists = result.source().pipelineContext().downstreamOfType(result.source(), TopN.class);
+          if (topNExists == null || topNExists.size() == 0) {
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: And need threshold limit so we only do this if we have more than X number of time series..

Comment (gskchaitanya) [reply core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: We can always run it in parallel mode even if it's just a single timeseries

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java (hunk 1 above)
Comment (gskchaitanya) [reply core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: good catch

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java (hunk 4 above)
Comment (gskchaitanya) [reply core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: wasStatus is an AtomicBoolean, so don't have to make it volatile

File: core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java (hunk 3 above)
Comment (gskchaitanya) [reply core/src/main/java/net/opentsdb/query/execution/serdes/JsonV3QuerySerdes.java]: incase of sequential, one thread shouldn't have an impact on the locking.

**Pull request merged**